internal/credential/vault/repository_credential_library.go

```diff
@@ -452,23 +457,243 @@ func (r *Repository) DeleteCredentialLibrary(ctx context.Context, projectId stri
+	slices.SortFunc(libs, func(i, j credential.Library) int {
+		return j.GetCreateTime().AsTime().Compare(i.GetCreateTime().AsTime())
+	})
+
+	return libs, transactionTimestamp, nil
+}
+
+// ListLibrariesRefresh returns a slice of credential libraries
+// for the store ID. Supports the following options:
+//   - credential.WithLimit
+//   - credential.WithStartPageAfterItem
+func (r *Repository) ListLibrariesRefresh(ctx context.Context, storeId string, updatedAfter time.Time, opt ...credential.Option) ([]credential.Library, time.Time, error) {
+	const op = "vault.(Repository).ListLibrariesRefresh"
+	switch {
+	case storeId == "":
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing credential store ID")
+	case updatedAfter.IsZero():
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing updated after time")
+	}
+	opts, err := credential.GetOpts(opt...)
+	if err != nil {
+		return nil, time.Time{}, errors.Wrap(ctx, err, op)
+	}
+	limit := r.defaultLimit
+	if opts.WithLimit != 0 {
+		// non-zero signals an override of the default limit for the repo.
+		limit = opts.WithLimit
+	}
+
+	query := fmt.Sprintf(listLibrariesRefreshTemplate, limit)
+	args := []any{
+		sql.Named("store_id", storeId),
+		sql.Named("updated_after_time", timestamp.New(updatedAfter)),
+	}
+	if opts.WithStartPageAfterItem != nil {
+		query = fmt.Sprintf(listLibrariesRefreshPageTemplate, limit)
+		args = append(args,
+			sql.Named("last_item_update_time", opts.WithStartPageAfterItem.GetUpdateTime()),
+			sql.Named("last_item_id", opts.WithStartPageAfterItem.GetPublicId()),
+		)
+	}
+
+	libs, transactionTimestamp, err := r.queryLibraries(ctx, query, args)
+	if err != nil {
+		return nil, time.Time{}, errors.Wrap(ctx, err, op)
+	}
+
+	// Sort final slice to ensure correct ordering.
+	// We sort by update time descending (most recently updated first).
+	slices.SortFunc(libs, func(i, j credential.Library) int {
+		return j.GetUpdateTime().AsTime().Compare(i.GetUpdateTime().AsTime())
+	})
+
+	return libs, transactionTimestamp, nil
+}
+
+func (r *Repository) queryLibraries(ctx context.Context, query string, args []any) ([]credential.Library, time.Time, error) {
+	const op = "vault.(Repository).queryLibraries"
+
+	var libs []credential.Library
+	var transactionTimestamp time.Time
+	if _, err := r.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{}, func(rd db.Reader, w db.Writer) error {
+		rows, err := rd.Query(ctx, query, args)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var (
+				publicId, storeId, projectId, name, description,
+				vaultPath, credentialType, httpMethod, httpRequestBody,
+				username, keyType, ttl, keyId, criticalOptions,
+				extensions, additionalValidPrincipals sql.NullString
+				createTime, updateTime *timestamp.Timestamp
+				version, keyBits       sql.NullInt64
+				typ                    string
+			)
```

Comment:
A struct would perhaps make the mapping from column to field safer since it would use the field name to map to the query result rather than just the position in the query 👍🏻. I'll see if I can do the same thing for credentials while I'm at it (in a separate PR).

Reply:
Done